File: problems/leetcode/0200/200/200.cpp
// 200. Number of Islands
// https://leetcode.com/problems/number-of-islands/

using namespace std;

#include <vector>

class Solution {
public:
  int m, n;
  vector<vector<bool>> visited;
  void dfs(vector<vector<char>> &grid, int i, int j) {
    visited[i][j] = true;
    if (i + 1 < m && grid[i + 1][j] == '1' && !visited[i + 1][j]) {
      dfs(grid, i + 1, j);
    }
    if (i - 1 >= 0 && grid[i - 1][j] == '1' && !visited[i - 1][j]) {
      dfs(grid, i - 1, j);
    }
    if (j + 1 < n && grid[i][j + 1] == '1' && !visited[i][j + 1]) {
      dfs(grid, i, j + 1);
    }
    if (j - 1 >= 0 && grid[i][j - 1] == '1' && !visited[i][j - 1]) {
      dfs(grid, i, j - 1);
    }
  }

  int numIslands(vector<vector<char>> &grid) {
    m = grid.size();
    n = grid[0].size();
    int answer = 0;
    vector<vector<bool>> _visited(m, vector<bool>(n, false));
    visited = _visited;

    for (int i = 0; i < m; i++) {
      for (int j = 0; j < n; j++) {
        if (grid[i][j] == '1' && !visited[i][j]) {
          dfs(grid, i, j);
          answer++;
        }
      }
    }

    return answer;
  }
};

File: problems/leetcode/0200/200/200.py
# 200. Number of Islands
# https://leetcode.com/problems/number-of-islands/description

from collections import deque
from typing import *

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m, n = len(grid), len(grid[0])
        visited = set()
        islands = 0

        for i in range(m):
            for j in range(n):
                if grid[i][j] == "1" and (i, j) not in visited:
                    islands += 1

                    q = deque()
                    q.append((i, j))
                    visited.add((i, j))

                    while q:
                        r, c = q.popleft()

                        for dr, dc in [(1, 0), (-1, 0), (0, 1), (0, -1)]:
                            nextR, nextC = r + dr, c + dc
                            if 0 <= nextR < m and 0 <= nextC < n and grid[nextR][nextC] == "1" and (nextR, nextC) not in visited:
                                visited.add((nextR, nextC))
                                q.append((nextR, nextC))

        return islands
